### PR TITLE
Fix insertNodes when replacing content adjacent to an inline element

### LINKED ISSUE #3136: Bug: Calling `insertNodes` when a text node is selected, followed by an inline element, causes the nodes to be inserted inside or after the element
Lexical version: 0.5.0

## Steps To Reproduce

1. Create a paragraph with a text node, followed by an inline element, e.g. a link element
2. Fully-select the text node
3. Call `insertNodes` providing some content to replace the text node with

e.g.

```ts
await editor.update(() => {
  const root = $getRoot();

  const paragraph = $createParagraphNode();
  root.append(paragraph);

  const text = $createTextNode('Existing text...');
  paragraph.append(text);

  setAnchorPoint({
    key: text.getKey(),
    offset: 0,
    type: 'text',
  });
  setFocusPoint({
    key: text.getKey(),
    offset: 'Existing text...'.length,
    type: 'text',
  });

  const selection = $getSelection();

  if (!$isRangeSelection(selection)) {
    return;
  }
  selection.insertNodes([$createTextNode('foo')]);
});
```

## The current behavior

Given the above example, the following (simplified) HTML will be produced:

```html
<p><a href=\"https://"><span>link</span></a><span>foo</span></p>
```

Notice that the inserted text has been inserted _after_ the sibling inline element, despite the fact that we selected text _before_ the inline element.

## The expected behavior

The inserted nodes should replace the selected content. That is, they should be inserted before the sibling inline element, i.e.

```html
<p><span>foo</span><a href=\"https://"><span>link</span></a></p>
```

## Analysis

This appears to occur because at the start of `RangeSelection.insertNodes` we collapse the selection by removing its content:

https://github.com/facebook/lexical/blob/06e347070ade5ca07ac7d27ef45a544d4f8efb2a/packages/lexical/src/LexicalSelection.ts#L1199-L1201

As a result of this, the selection will be moved to the remaining child, specifically it will be put on the link _element_ at offset 0.

Following that, we try to resolve the target node to insert content _after_ with the following steps:

https://github.com/facebook/lexical/blob/06e347070ade5ca07ac7d27ef45a544d4f8efb2a/packages/lexical/src/LexicalSelection.ts#L1208-L1218

This makes sense for a paragraph that has since been emptied, but in our case `placementNode` will be `null` and we will end up setting the `target` to the inline element (i.e. the link in this example).

From here, the behavior depends on the type of content we are trying to insert:

- For a text node, for example, we'll end up inserting it _after_ the inline element: https://github.com/facebook/lexical/blob/06e347070ade5ca07ac7d27ef45a544d4f8efb2a/packages/lexical/src/LexicalSelection.ts#L1390-L1391
- For a block element, we'll end up trying to merge its contents and insert its content inside the inline element

### PR DESCRIPTION
**Pull request by birtles**
Fixes #3136.

### CODE REVIEW

**Comment by birtles** [comment]
Vercel gives me a 404 when I try to view the playground failure so I'm not sure what's going on there

**Comment by trueadm** [comment]
Please could you add an e2e test for this given the core nature of insert nodes? Thank you

**Comment by birtles** [comment]
> Please could you add an e2e test for this given the core nature of insert nodes? Thank you

Sure, it's the start of a long weekend here but I'll get onto it next week.

File: packages/lexical/src/LexicalSelection.ts [hunk 1]
```diff
@@ -1197,7 +1197,27 @@ export class RangeSelection implements BaseSelection {
   insertNodes(nodes: Array<LexicalNode>, selectStart?: boolean): boolean {
     // If there is a range selected remove the text in it
     if (!this.isCollapsed()) {
+      const selectionEnd = this.isBackward() ? this.anchor : this.focus;
+      const nextSiblingKey = selectionEnd.getNode().getNextSibling()?.getKey();
```
Comment (acywatson) [inline packages/lexical/src/LexicalSelection.ts:1201]: You can't use safe navigation in this repo - this is why the integrity check and vercel build are failing.

**Comment by birtles** [comment]
> Please could you add an e2e test for this given the core nature of insert nodes? Thank you

E2E test added.

**Comment by birtles** [comment]
Actually, I think this same situation might also arise when removing content from the end of a block.

**Comment by birtles** [comment]
(Test failures appear to be pre-existing flakiness that I can reproduce without this patch applied.)

**Comment by birtles** [comment]
> Actually, I think this same situation might also arise when removing content from the end of a block.

Updated the PR to handle the end of block case too.

**Review by fantactuka — APPROVED** [review]

**Pull request merged**